File: EduAnalytics/frontend/src/utils/gradeConverter.js
/**
 * Grade Conversion Utility for Frontend
 * Converts between marks (0-100) and letter grades
 * Matches backend grading scheme exactly
 */

export const GRADE_SCHEME = {
  'O': { min: 91, max: 100, description: 'Outstanding' },
  'A+': { min: 81, max: 90, description: 'Excellent' },
  'A': { min: 71, max: 80, description: 'Very Good' },
  'B+': { min: 61, max: 70, description: 'Good' },
  'B': { min: 56, max: 60, description: 'Average' },
  'C': { min: 50, max: 55, description: 'Satisfactory' },
  'RA': { min: 0, max: 49, description: 'Re-Appear' }
};

/**
 * Convert numeric marks to letter grade
 * @param {number} marks - Marks between 0-100
 * @returns {string} Letter grade (O, A+, A, B+, B, C, RA)
 */
export const getGradeFromMarks = (marks) => {
  if (marks === null || marks === undefined || marks === '') {
    return '';
  }

  const numMarks = parseFloat(marks);
  
  if (isNaN(numMarks)) {
    return '';
  }

  if (numMarks >= 91) return 'O';
  if (numMarks >= 81) return 'A+';
  if (numMarks >= 71) return 'A';
  if (numMarks >= 61) return 'B+';
  if (numMarks >= 56) return 'B';
  if (numMarks >= 50) return 'C';
  return 'RA';
};

/**
 * Get grade description
 * @param {string} grade - Letter grade
 * @returns {string} Full description
 */
export const getGradeDescription = (grade) => {
  if (!grade || !GRADE_SCHEME[grade]) {
    return '';
  }
  const scheme = GRADE_SCHEME[grade];
  return `${grade} (${scheme.min}-${scheme.max}) - ${scheme.description}`;
};

/**
 * Validate if a grade is passing (not RA)
 * @param {string} grade - Letter grade
 * @returns {boolean} True if passing, false if RA or empty
 */
export const isPassGrade = (grade) => {
  if (!grade) return false;
  return grade.toUpperCase() !== 'RA';
};

/**
 * Get color for grade display
 * @param {string} grade - Letter grade
 * @returns {string} Tailwind color class
 */
export const getGradeColor = (grade) => {
  if (!grade) return 'text-gray-500';
  
  const gradeUpper = grade.toUpperCase();
  switch (gradeUpper) {
    case 'O':
      return 'text-green-600 font-bold';
    case 'A+':
      return 'text-green-500 font-bold';
    case 'A':
      return 'text-blue-600 font-semibold';
    case 'B+':
      return 'text-blue-500 font-semibold';
    case 'B':
      return 'text-yellow-600 font-semibold';
    case 'C':
      return 'text-orange-500';
    case 'RA':
      return 'text-red-600 font-bold';
    default:
      return 'text-gray-500';
  }
};

/**
 * Get background color for grade badge
 * @param {string} grade - Letter grade
 * @returns {string} Tailwind bg color class
 */
export const getGradeBgColor = (grade) => {
  if (!grade) return 'bg-gray-100';
  
  const gradeUpper = grade.toUpperCase();
  switch (gradeUpper) {
    case 'O':
      return 'bg-green-100';
    case 'A+':
      return 'bg-green-50';
    case 'A':
      return 'bg-blue-100';
    case 'B+':
      return 'bg-blue-50';
    case 'B':
      return 'bg-yellow-100';
    case 'C':
      return 'bg-orange-100';
    case 'RA':
      return 'bg-red-100';
    default:
      return 'bg-gray-100';
  }
};

/**
 * Get all valid grades
 * @returns {array} Array of grade strings
 */
export const getAllGrades = () => {
  return Object.keys(GRADE_SCHEME);
};

File: EduAnalytics/backend/app/utils/grade_converter.py
"""
Grade Conversion Utilities for Semester Examination
Converts between grades and numeric values according to grading scheme
"""

# Grading Scheme Mapping
GRADE_RANGES = {
    'O': (91, 100),      # Outstanding
    'A+': (81, 90),      # Excellent
    'A': (71, 80),       # Very Good
    'B+': (61, 70),      # Good
    'B': (56, 60),       # Average
    'C': (50, 55),       # Satisfactory
    'RA': (0, 49)        # Re-Appear
}

GRADE_TO_MIDPOINT = {
    'O': 95.5,      # (91+100)/2
    'A+': 85.5,     # (81+90)/2
    'A': 75.5,      # (71+80)/2
    'B+': 65.5,     # (61+70)/2
    'B': 58,        # (56+60)/2
    'C': 52.5,      # (50+55)/2
    'RA': 24.5      # (0+49)/2
}

def get_grade_from_marks(marks: float) -> str:
    """
    Convert numeric marks (0-100) to letter grade
    
    Args:
        marks: Float value between 0-100
        
    Returns:
        Letter grade: O, A+, A, B+, B, C, or RA
    """
    if marks is None:
        return None
        
    marks = float(marks)
    
    if marks >= 91:
        return 'O'
    elif marks >= 81:
        return 'A+'
    elif marks >= 71:
        return 'A'
    elif marks >= 61:
        return 'B+'
    elif marks >= 56:
        return 'B'
    elif marks >= 50:
        return 'C'
    else:
        return 'RA'


def get_numeric_from_grade(grade: str) -> float:
    """
    Convert letter grade to numeric midpoint value
    Used for calculations when semester results are published
    
    Args:
        grade: Letter grade (O, A+, A, B+, B, C, RA)
        
    Returns:
        Midpoint numeric value for the grade range
    """
    if grade is None:
        return None
        
    return GRADE_TO_MIDPOINT.get(grade.upper(), None)


def validate_grade(grade: str) -> bool:
    """
    Validate if grade is in accepted format
    
    Args:
        grade: Letter grade to validate
        
    Returns:
        True if valid, False otherwise
    """
    if grade is None:
        return True  # None is valid (not yet published)
        
    return grade.upper() in GRADE_RANGES or grade.upper() in GRADE_TO_MIDPOINT


def get_grade_description(grade: str) -> str:
    """
    Get full description of a grade
    
    Args:
        grade: Letter grade
        
    Returns:
        Description string
    """
    descriptions = {
        'O': 'Outstanding (91-100)',
        'A+': 'Excellent (81-90)',
        'A': 'Very Good (71-80)',
        'B+': 'Good (61-70)',
        'B': 'Average (56-60)',
        'C': 'Satisfactory (50-55)',
        'RA': 'Re-Appear (<50)'
    }
    return descriptions.get(grade.upper(), 'Unknown')


def is_pass_grade(grade: str) -> bool:
    """
    Check if a grade is a passing grade (not RA)
    
    Args:
        grade: Letter grade
        
    Returns:
        True if passing, False if RA or None
    """
    if grade is None:
        return None
    return grade.upper() != 'RA'


def get_ca_pass_status(ca_total: float) -> bool:
    """
    Check if CA marks qualify as pass (>= 30 out of 60)
    CA is average of CA1, CA2, CA3 which total to 300 marks
    Pass requirement: average >= 30/60 = 50%
    
    Args:
        ca_total: Sum of (CA1 + CA2 + CA3) / 3 (average CA marks)
        
    Returns:
        True if CA average >= 30, False otherwise
    """
    if ca_total is None:
        return False
    return float(ca_total) >= 30


# Grade distribution helper
def get_grade_distribution_counts(grades_list: list) -> dict:
    """
    Count distribution of grades
    
    Args:
        grades_list: List of grade strings
        
    Returns:
        Dictionary with counts: {'O': n, 'A+': n, ...}
    """
    distribution = {
        'O': 0,
        'A+': 0,
        'A': 0,
        'B+': 0,
        'B': 0,
        'C': 0,
        'RA': 0
    }
    
    for grade in grades_list:
        if grade and grade.upper() in distribution:
            distribution[grade.upper()] += 1
    
    return distribution
